corehq/apps/domain/models.py

```diff
@@ -1227,3 +1229,46 @@ def as_dict(self):
             'deactivate_url': self.deactivate_url(),
             'activate_url': self.activate_url(),
         }
+
+
+class DomainAuditRecordEntry(models.Model):
+    id = models.UUIDField(unique=True, default=uuid.uuid4, primary_key=True)
+    domain = models.TextField(unique=True, db_index=True)
+    calculations = JSONField(default=dict)
+
+    @classmethod
+    def update_calculations(cls, domain, model, method=None):
+        obj, is_new = cls.objects.get_or_create(domain=domain)
+        if is_new:
+            calculations = {
+                "cp_n_downloads_custom_exports": 0,
+                "cp_n_viewed_ucr_reports": 0,
+                "cp_n_viewed_non_ucr_reports": 0,
+                "cp_n_reports_created": 0,
+                "cp_n_reports_edited": 0,
+                "cp_n_saved_scheduled_reports": 0,
+                "cp_n_click_app_deploy": 0,
+                "cp_n_form_builder_entered": 0,
+                "cp_n_saved_app_changes": 0
+            }
+        else:
+            calculations = obj.calculations
+
+        config = {
+            ('retrieve_download', None): "cp_n_downloads_custom_exports",
+            ('ConfigurableReportView', None): "cp_n_viewed_ucr_reports",
+            ('ProjectReportDispatcher', None): "cp_n_viewed_non_ucr_reports",
+            ('ReportConfiguration', 'create'): "cp_n_reports_created",
+            ('ReportConfiguration', 'update'): "cp_n_reports_edited",
+            ('ReportNotification', None): "cp_n_saved_scheduled_reports",
+            ('release_build', None): "cp_n_click_app_deploy",
+            ('form_source', None): "cp_n_form_builder_entered",
+            ('patch_xform', None): "cp_n_saved_app_changes",
+            ('edit_module_attr', None): "cp_n_saved_app_changes",
+            ('edit_app_attr', None): "cp_n_saved_app_changes"
+        }
+
+        property_to_update = config.get((model, method))
+        calculations[property_to_update] += 1
```

Comment:
This isn't atomic, because you could encounter a race condition with non-atomic incrementing.  If you have a column field for each calculated property, then you an do an atomic increment by using django F statements.

https://stackoverflow.com/questions/1598932/atomic-increment-of-a-counter-in-django?answertab=votes#tab-top

Reply:
:+1: